### PR TITLE
Streams support

### PR DESCRIPTION
**Pull request by stleary**
**What problem does this code solve?**
JSONObject and JSONArray should support streams for functional programming. 
There are multiple ways this could be added. If you know of a better way, please post your comments to this PR.

This change requires Java 8. Going forward, Java 6 and Java 7 library builds are no longer supported. The last release for Java 6/7 is 20230618. Users who still need Java 6/7 support should open an issue in this repo.

**Risks**
Low

**Changes to the API?**
New API methods were added

**Will this require a new release?**
Yes

**Should the documentation be updated?**
No

**Does it break the unit tests?**
No. New unit tests were added

**Was any code refactored in this commit?**

* gradlew permissions were updated
* ParserConfiguration warnings were addressed

**Review status**
code review action items pending

### CODE REVIEW

**Review by johnjaylward — CHANGES_REQUESTED** [review]
Overall this looks like a good start, However I'd personally like to see a few more test cases around getting data out and `null` handling in both the `JSONObject` and `JSONArray` streams.

Looking at #673 and #674, I think we should look towards always returning valid JSON value types from the streams. It is something we missed out on with the iterators and prevents people from having a consistent interface.

We should also be sure that the streams themselves (specifically the ones provided as part of the API, not the follow-on calls as part of the tests) do not throw exceptions by using `opt` operations wherever possible.

File: src/main/java/org/json/JSONArray.java [hunk 1]
```diff
@@ -224,6 +226,14 @@ public JSONArray(int initialCapacity) throws JSONException {
     	this.myArrayList = new ArrayList<Object>(initialCapacity);
     }
 
+    /**
+     * Stream support - requires java 8
+     * @return Stream of array elements
+     */
+    public Stream<Object> stream() {
+        return IntStream.range(0, length()).mapToObj(this::get);
```
[the remark this answers is not in the document]
Comment (stleary) [reply src/main/java/org/json/JSONArray.java:234]: > I think this should not be `this::get` and instead should be `this::opt`. Stream should not throw exceptions, which `get` will do if a value is `null`.
> 
> We should also be sure to include at least one test case that has `null` in the array.

Seems to be working as is, but not sure why. Test cases are updated, let me know if this code or the tests should still be changed.

File: src/main/java/org/json/JSONObject.java [hunk 2]
```diff
@@ -81,7 +74,13 @@
  * @author JSON.org
  * @version 2016-08-15
  */
-public class JSONObject {
+public class JSONObject implements Iterable<String> {
```
[the remark this answers is not in the document]
Comment (stleary) [reply src/main/java/org/json/JSONObject.java:77]: Implementation now streams entrySet() without Iterable<>.

**Comment by stleary** [comment]
> I'm trying to add review notes here, but it seems that GitHub is having issues.
> 
> I think for this implementation we should also consider #673 and #674.

Can this be done in a subsequent commit? The other issues seem out of scope for this initial implementation.

File: src/test/java/org/json/junit/JSONArrayTest.java [hunk 3]
```diff
@@ -66,6 +61,194 @@ public class JSONArrayTest {
...
+    }
+
+    /**
+     * 1.3.2 Stream null JSONArray
+     */
+    @Test
+    public void testStream_1_3_2() {
+        JSONArray jsonArray = null;
+
+        assertThrows(NullPointerException.class, () -> {
```
Comment (johnjaylward) [inline src/test/java/org/json/junit/JSONArrayTest.java:223]: I don't think this test is adding anything here. It seems obvious that an NPE would be thrown in this case.

File: src/test/java/org/json/junit/JSONObjectTest.java [hunk 4]
```diff
@@ -72,6 +67,253 @@ public class JSONObjectTest {
...
+    }
+
+    /**
+     * Test Case 2.3.2: stream a null JSONObject
+     */
+    @Test
+    public void test_2_3_2() {
+        JSONObject jsonObject = null;
+
+        assertThrows(NullPointerException.class, () -> {
```
Comment (johnjaylward) [inline src/test/java/org/json/junit/JSONObjectTest.java:282]: I don't think this test is adding anything here. It seems obvious that an NPE would be thrown in this case.

File: src/main/java/org/json/JSONArray.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnjaylward) [reply src/main/java/org/json/JSONArray.java:234]: I stepped through the code and that Parser is converting `null` to `JSONObject.NULL`, which is expected.

Let's add these 2 additional tests, which are very similar, but add different ways of composing the data instead of just using the parser. This second test shows the issue I was worried about:

```java
@Test
    public void testStream_ComposeFromArray() {
        @SuppressWarnings("boxing")
        Object[] expectedList = new Object[] { 1, 2, 3, true, false, null, "abc", "def", new JSONObject("{ \"a\": \"b\" }"), new JSONArray()};
        JSONArray jsonArray = new JSONArray(expectedList);
        
        List<Object> actualList = jsonArray.stream()
                .collect(Collectors.toList());
        System.out.println(actualList);
        assertEquals(actualList.size(), expectedList.length);
        for (int i = 0; i < actualList.size(); ++i) {
            Object actual = actualList.get(i);
            Object expected = expectedList[i];
            if (actual instanceof JSONObject) {
                assertTrue("Json Object test",((JSONObject) actual).similar(expected));
            } else if (actual instanceof JSONArray) {
                assertTrue("Json Array test", ((JSONArray) actual).similar(expected));
            } else if (JSONObject.NULL.equals(expected)) {
                assertTrue("Null test", JSONObject.NULL.equals(actual));
            } else {
                assertEquals("value test: "+expected, expected, actual);
            }
        }
    }
```

```java
    @Test
    public void testStream_ComposeFromCode() {
        JSONArray jsonArray = new JSONArray();
        jsonArray.put(1).put(2).put(3)
            .put(true).put(false)
            .put((Object)null)
            .put("abc").put("def")
            .put(new JSONObject("{ \"a\": \"b\" }"))
            .put(new JSONArray())
        ;
        @SuppressWarnings("boxing")
        Object[] expectedList = new Object[] { 1, 2, 3, true, false, null, "abc", "def", new JSONObject("{ \"a\": \"b\" }"), new JSONArray()};
        
        List<Object> actualList = jsonArray.stream()
                .collect(Collectors.toList());
        System.out.println(actualList);
        assertEquals(actualList.size(), expectedList.length);
        for (int i = 0; i < actualList.size(); ++i) {
            Object actual = actualList.get(i);
            Object expected = expectedList[i];
            if (actual instanceof JSONObject) {
                assertTrue("Json Object test",((JSONObject) actual).similar(expected));
            } else if (actual instanceof JSONArray) {
                assertTrue("Json Array test", ((JSONArray) actual).similar(expected));
            } else if (JSONObject.NULL.equals(expected)) {
                assertTrue("Null test", JSONObject.NULL.equals(actual));
            } else {
                assertEquals("value test: "+expected, expected, actual);
            }
        }
    }
```

[the remark this answers is not in the document]
Comment (stleary) [reply src/main/java/org/json/JSONArray.java]: Added tests

File: src/test/java/org/json/junit/JSONArrayTest.java [hunk 5]
```diff
@@ -66,6 +68,42 @@ public class JSONArrayTest {
...
+                "    {\n" +
+                "      \"id\": 3,\n" +
+                "      \"name\": \"Carol\",\n" +
+                "      \"department\": \"Marketing\"\n" +
+                "    }\n" +
+                "  ]";
+        JSONArray jsonArray = new JSONArray(data);
+
+        // Get employees from the engineering dept
+        List<String> actualList = StreamSupport.stream(jsonArray.spliterator(), false)
```
[the remark this answers is not in the document]
Comment (stleary) [reply src/test/java/org/json/junit/JSONArrayTest.java]: Removed

File: src/test/java/org/json/junit/JSONArrayTest.java (hunk 3 above)
Comment (stleary) [reply src/test/java/org/json/junit/JSONArrayTest.java]: Removed

File: src/test/java/org/json/junit/JSONObjectTest.java (hunk 4 above)
Comment (stleary) [reply src/test/java/org/json/junit/JSONObjectTest.java]: Removed

**Comment by stleary** [comment]
Added the new unit tests

**Comment by stleary** [comment]
Closing due to lack of activity.
